core/lib/state/src/storage_factory/mod.rs

```diff
@@ -119,6 +119,58 @@ impl CommonStorage<'static> {
         Ok(Some(rocksdb.into()))
     }
 
+    /// Returns a [`ReadStorage`] implementation backed by [`RocksDBWithMemory`]
+    ///     and the result of [`RocksdbStorage::l1_batch_number`].
+    /// RocksDB isn't caught up, instead batch diffs are used to fill the gap.
+    /// Note, as long as `RocksDBWithMemory` objects are held only for `l1_batch_number >= X`,
+    /// then it's ok if RocksDB is caught up in parallel for L1 batch number up to X
+    /// because affected storage diff will be read from memory anyway.
+    ///
+    /// # Errors
+    ///
+    /// Propagates RocksDB errors.
+    pub async fn rocksdb_with_memory(
+        rocksdb: RocksDB<StateKeeperColumnFamily>,
+        batch_diffs: &BatchDiffs,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<(Self, L1BatchNumber)> {
+        tracing::debug!("Catching up RocksDB synchronously");
+        let rocksdb_builder = RocksdbStorageBuilder::from_rocksdb(rocksdb);
+
+        let rocksdb = rocksdb_builder.skip_synchronize();
+        let rocksdb_l1_batch_number = rocksdb
+            .l1_batch_number()
+            .await
+            .ok_or_else(|| anyhow::anyhow!("No L1 batches available in Postgres"))?;
```

Comment:
- Bikeshedding: It could make sense at this point to accept `RocksdbStorage` as an argument and allow to create it from `RocksDB<_>` (e.g., via `From<_>`). Initially, the builder pattern was used to prevent uninitialized `RocksdbStorage`, but it's evidently no longer working with these changes. This would allow to not return `L1BatchNumber` and not to add `skip_synchronize()`.
- Nit: It's possible to use `.context(_)` instead of `.ok_or_else()` here.